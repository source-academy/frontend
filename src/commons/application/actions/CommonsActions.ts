import { Router } from '@remix-run/router';
import { action } from 'typesafe-actions'; // EDITED

import { LOG_OUT, UPDATE_REACT_ROUTER } from '../types/CommonsTypes';

export const logOut = () => action(LOG_OUT);
export const updateReactRouter = (updatedRouter: Router) =>
  action(UPDATE_REACT_ROUTER, updatedRouter);
