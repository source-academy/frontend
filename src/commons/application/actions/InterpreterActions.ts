import { SourceError, Value } from 'js-slang/dist/types';
import { action } from 'typesafe-actions';

import { WorkspaceLocation } from '../../workspace/WorkspaceTypes';
import {
  BEGIN_DEBUG_PAUSE,
  BEGIN_INTERRUPT_EXECUTION,
  DEBUG_RESET,
  DEBUG_RESUME,
  END_DEBUG_PAUSE,
  END_INTERRUPT_EXECUTION,
  EVAL_INTERPRETER_ERROR,
  EVAL_INTERPRETER_SUCCESS,
  EVAL_TESTCASE_FAILURE,
  EVAL_TESTCASE_SUCCESS,
  HANDLE_CONSOLE_LOG
} from '../types/InterpreterTypes';

export const handleConsoleLog = (workspaceLocation: WorkspaceLocation, ...logString: string[]) =>
  action(HANDLE_CONSOLE_LOG, { logString, workspaceLocation });

export const evalInterpreterSuccess = (value: Value, workspaceLocation: WorkspaceLocation) =>
  action(EVAL_INTERPRETER_SUCCESS, { type: 'result', value, workspaceLocation });

export const evalTestcaseSuccess = (
  value: Value,
  workspaceLocation: WorkspaceLocation,
  index: number
) => action(EVAL_TESTCASE_SUCCESS, { type: 'result', value, workspaceLocation, index });

export const evalTestcaseFailure = (
  value: Value,
  workspaceLocation: WorkspaceLocation,
  index: number
) => action(EVAL_TESTCASE_FAILURE, { type: 'errors', value, workspaceLocation, index });

export const evalInterpreterError = (errors: SourceError[], workspaceLocation: WorkspaceLocation) =>
  action(EVAL_INTERPRETER_ERROR, { type: 'errors', errors, workspaceLocation });

export const beginInterruptExecution = (workspaceLocation: WorkspaceLocation) =>
  action(BEGIN_INTERRUPT_EXECUTION, { workspaceLocation });

export const endInterruptExecution = (workspaceLocation: WorkspaceLocation) =>
  action(END_INTERRUPT_EXECUTION, { workspaceLocation });

export const beginDebuggerPause = (workspaceLocation: WorkspaceLocation) =>
  action(BEGIN_DEBUG_PAUSE, { workspaceLocation });

export const endDebuggerPause = (workspaceLocation: WorkspaceLocation) =>
  action(END_DEBUG_PAUSE, { workspaceLocation });

export const debuggerResume = (workspaceLocation: WorkspaceLocation) =>
  action(DEBUG_RESUME, { workspaceLocation });

export const debuggerReset = (workspaceLocation: WorkspaceLocation) =>
  action(DEBUG_RESET, { workspaceLocation });
