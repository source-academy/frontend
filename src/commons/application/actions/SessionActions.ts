import { action } from 'typesafe-actions'; // EDITED

import { Grading, GradingOverview } from '../../../features/grading/GradingTypes';
import {
  Assessment,
  AssessmentConfiguration,
  AssessmentOverview,
  ContestEntry
} from '../../assessment/AssessmentTypes';
import { MissionRepoData } from '../../githubAssessments/GitHubMissionTypes';
import {
  Notification,
  NotificationFilterFunction
} from '../../notificationBadge/NotificationBadgeTypes';
import { generateOctokitInstance } from '../../utils/GitHubPersistenceHelper';
import { Role } from '../ApplicationTypes';
import {
  ACKNOWLEDGE_NOTIFICATIONS,
  AdminPanelCourseRegistration,
  CourseRegistration,
  DELETE_ASSESSMENT_CONFIG,
  DELETE_USER_COURSE_REGISTRATION,
  FETCH_ADMIN_PANEL_COURSE_REGISTRATIONS,
  FETCH_ALL_USER_XP,
  FETCH_ASSESSMENT,
  FETCH_ASSESSMENT_ADMIN,
  FETCH_ASSESSMENT_CONFIGS,
  FETCH_ASSESSMENT_OVERVIEWS,
  FETCH_AUTH,
  FETCH_COURSE_CONFIG,
  FETCH_GRADING,
  FETCH_GRADING_OVERVIEWS,
  FETCH_NOTIFICATIONS,
  FETCH_TOTAL_XP,
  FETCH_TOTAL_XP_ADMIN,
  FETCH_USER_AND_COURSE,
  LOGIN,
  LOGIN_GITHUB,
  LOGOUT_GITHUB,
  LOGOUT_GOOGLE,
  REAUTOGRADE_ANSWER,
  REAUTOGRADE_SUBMISSION,
  REMOVE_GITHUB_OCTOKIT_OBJECT_AND_ACCESS_TOKEN,
  SET_ADMIN_PANEL_COURSE_REGISTRATIONS,
  SET_ASSESSMENT_CONFIGURATIONS,
  SET_COURSE_CONFIGURATION,
  SET_COURSE_REGISTRATION,
  SET_GITHUB_ACCESS_TOKEN,
  SET_GITHUB_ASSESSMENT,
  SET_GITHUB_OCTOKIT_OBJECT,
  SET_GOOGLE_USER,
  SET_TOKENS,
  SET_USER,
  SUBMIT_ANSWER,
  SUBMIT_ASSESSMENT,
  SUBMIT_GRADING,
  SUBMIT_GRADING_AND_CONTINUE,
  Tokens,
  UNSUBMIT_SUBMISSION,
  UPDATE_ALL_USER_XP,
  UPDATE_ASSESSMENT,
  UPDATE_ASSESSMENT_CONFIGS,
  UPDATE_ASSESSMENT_OVERVIEWS,
  UPDATE_COURSE_CONFIG,
  UPDATE_COURSE_RESEARCH_AGREEMENT,
  UPDATE_GRADING,
  UPDATE_GRADING_OVERVIEWS,
  UPDATE_LATEST_VIEWED_COURSE,
  UPDATE_NOTIFICATIONS,
  UPDATE_TOTAL_XP,
  UPDATE_USER_ROLE,
  UpdateCourseConfiguration,
  User
} from '../types/SessionTypes';

export const fetchAuth = (code: string, providerId?: string) =>
  action(FETCH_AUTH, { code, providerId });

export const fetchUserAndCourse = () => action(FETCH_USER_AND_COURSE);

export const fetchCourseConfig = () => action(FETCH_COURSE_CONFIG);

export const fetchAssessment = (assessmentId: number) => action(FETCH_ASSESSMENT, assessmentId);

export const fetchAssessmentAdmin = (assessmentId: number, courseRegId: number) =>
  action(FETCH_ASSESSMENT_ADMIN, { assessmentId, courseRegId });

export const fetchAssessmentOverviews = () => action(FETCH_ASSESSMENT_OVERVIEWS);

export const fetchTotalXp = () => action(FETCH_TOTAL_XP);

export const fetchTotalXpAdmin = (courseRegId: number) => action(FETCH_TOTAL_XP_ADMIN, courseRegId);

export const fetchAllUserXp = () => action(FETCH_ALL_USER_XP);

export const fetchGrading = (submissionId: number) => action(FETCH_GRADING, submissionId);

/**
 * @param filterToGroup - param when set to true, only shows submissions under the group
 * of the grader
 */
export const fetchGradingOverviews = (filterToGroup = true) =>
  action(FETCH_GRADING_OVERVIEWS, filterToGroup);

export const login = (providerId: string) => action(LOGIN, providerId);

export const logoutGoogle = () => action(LOGOUT_GOOGLE);

export const loginGitHub = () => action(LOGIN_GITHUB);

export const logoutGitHub = () => action(LOGOUT_GITHUB);

export const setTokens = ({ accessToken, refreshToken }: Tokens) =>
  action(SET_TOKENS, {
    accessToken,
    refreshToken
  });

export const setUser = (user: User) => action(SET_USER, user);

export const setCourseConfiguration = (courseConfiguration: UpdateCourseConfiguration) =>
  action(SET_COURSE_CONFIGURATION, courseConfiguration);

export const setCourseRegistration = (courseRegistration: Partial<CourseRegistration>) =>
  action(SET_COURSE_REGISTRATION, courseRegistration);

export const setAssessmentConfigurations = (assessmentConfigurations: AssessmentConfiguration[]) =>
  action(SET_ASSESSMENT_CONFIGURATIONS, assessmentConfigurations);

export const setAdminPanelCourseRegistrations = (
  courseRegistrations: AdminPanelCourseRegistration[]
) => action(SET_ADMIN_PANEL_COURSE_REGISTRATIONS, courseRegistrations);

export const setGoogleUser = (user?: string) => action(SET_GOOGLE_USER, user);

export const setGitHubAssessment = (missionRepoData: MissionRepoData) =>
  action(SET_GITHUB_ASSESSMENT, missionRepoData);

export const setGitHubOctokitObject = (authToken?: string) =>
  action(SET_GITHUB_OCTOKIT_OBJECT, generateOctokitInstance(authToken || ''));

export const setGitHubAccessToken = (authToken?: string) =>
  action(SET_GITHUB_ACCESS_TOKEN, authToken);

export const removeGitHubOctokitObjectAndAccessToken = () =>
  action(REMOVE_GITHUB_OCTOKIT_OBJECT_AND_ACCESS_TOKEN);

export const submitAnswer = (id: number, answer: string | number | ContestEntry[]) =>
  action(SUBMIT_ANSWER, {
    id,
    answer
  });

export const submitAssessment = (id: number) => action(SUBMIT_ASSESSMENT, id);

export const submitGrading = (
  submissionId: number,
  questionId: number,
  xpAdjustment: number = 0,
  comments?: string
) =>
  action(SUBMIT_GRADING, {
    submissionId,
    questionId,
    xpAdjustment,
    comments
  });

export const submitGradingAndContinue = (
  submissionId: number,
  questionId: number,
  xpAdjustment: number = 0,
  comments?: string
) =>
  action(SUBMIT_GRADING_AND_CONTINUE, {
    submissionId,
    questionId,
    xpAdjustment,
    comments
  });

export const reautogradeSubmission = (submissionId: number) =>
  action(REAUTOGRADE_SUBMISSION, submissionId);

export const reautogradeAnswer = (submissionId: number, questionId: number) =>
  action(REAUTOGRADE_ANSWER, { submissionId, questionId });

export const updateAssessmentOverviews = (overviews: AssessmentOverview[]) =>
  action(UPDATE_ASSESSMENT_OVERVIEWS, overviews);

export const updateTotalXp = (totalXp: number) => action(UPDATE_TOTAL_XP, totalXp);

export const updateAllUserXp = (allUserXp: string[][]) => action(UPDATE_ALL_USER_XP, allUserXp);

export const updateAssessment = (assessment: Assessment) => action(UPDATE_ASSESSMENT, assessment);

export const updateGradingOverviews = (overviews: GradingOverview[]) =>
  action(UPDATE_GRADING_OVERVIEWS, overviews);

/**
 * An extra id parameter is included here because of
 * no id for Grading.
 */
export const updateGrading = (submissionId: number, grading: Grading) =>
  action(UPDATE_GRADING, {
    submissionId,
    grading
  });

export const unsubmitSubmission = (submissionId: number) =>
  action(UNSUBMIT_SUBMISSION, {
    submissionId
  });

/**
 * Notification actions
 */

export const fetchNotifications = () => action(FETCH_NOTIFICATIONS);

export const acknowledgeNotifications = (withFilter?: NotificationFilterFunction) =>
  action(ACKNOWLEDGE_NOTIFICATIONS, {
    withFilter
  });

export const updateNotifications = (notifications: Notification[]) =>
  action(UPDATE_NOTIFICATIONS, notifications);

export const updateLatestViewedCourse = (courseId: number) =>
  action(UPDATE_LATEST_VIEWED_COURSE, { courseId });

export const updateCourseConfig = (courseConfiguration: UpdateCourseConfiguration) =>
  action(UPDATE_COURSE_CONFIG, courseConfiguration);

export const fetchAssessmentConfigs = () => action(FETCH_ASSESSMENT_CONFIGS);

export const updateAssessmentConfigs = (assessmentConfigs: AssessmentConfiguration[]) =>
  action(UPDATE_ASSESSMENT_CONFIGS, assessmentConfigs);

export const deleteAssessmentConfig = (assessmentConfig: AssessmentConfiguration) =>
  action(DELETE_ASSESSMENT_CONFIG, assessmentConfig);

export const fetchAdminPanelCourseRegistrations = () =>
  action(FETCH_ADMIN_PANEL_COURSE_REGISTRATIONS);

export const updateUserRole = (courseRegId: number, role: Role) =>
  action(UPDATE_USER_ROLE, { courseRegId, role });

export const deleteUserCourseRegistration = (courseRegId: number) =>
  action(DELETE_USER_COURSE_REGISTRATION, { courseRegId });

export const updateCourseResearchAgreement = (agreedToResearch: boolean) =>
  action(UPDATE_COURSE_RESEARCH_AGREEMENT, { agreedToResearch });
