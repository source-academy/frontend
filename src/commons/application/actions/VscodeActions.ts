import { createActions } from 'src/commons/redux/utils';

const VscodeActions = createActions('vscode', {
  setVscode: 0
});

// For compatibility with existing code (actions helper)
export default {
  ...VscodeActions
};
