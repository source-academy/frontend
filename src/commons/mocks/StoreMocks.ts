import _ from 'lodash';
import { DeepPartial, Store } from 'redux';
import mockStore from 'redux-mock-store';

import {
  defaultAcademy,
  defaultAchievement,
  defaultApplication,
  defaultDashboard,
  defaultFileSystem,
  defaultPlayground,
  defaultSession,
  defaultWorkspaceManager,
  OverallState
} from '../application/ApplicationTypes';

export function mockInitialStore(overrides?: DeepPartial<OverallState>): Store<OverallState> {
  const createStore = (mockStore as any)();
  const state: OverallState = {
    academy: defaultAcademy,
    achievement: defaultAchievement,
    application: defaultApplication,
    dashboard: defaultDashboard,
    playground: defaultPlayground,
    workspaces: defaultWorkspaceManager,
    session: defaultSession,
    fileSystem: defaultFileSystem
  };
  return createStore(_.merge(state, overrides));
}
