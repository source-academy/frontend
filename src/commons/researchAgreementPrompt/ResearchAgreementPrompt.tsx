import { Button, Classes, Dialog, H4, Intent } from '@blueprintjs/core';
import { useDispatch } from 'react-redux';

import { updateCourseResearchAgreement } from '../application/actions/SessionActions';
import Constants from '../utils/Constants';

const ResearchAgreementPrompt: React.FC = () => {
  const dispatch = useDispatch();

  return (
    <div className="research-prompt">
      <Dialog
        className={Classes.DARK}
        title="Agreement to Participate in Educational Research"
        canOutsideClickClose={false}
        canEscapeKeyClose={false}
        isCloseButtonShown={false}
        isOpen
      >
        <div className={Classes.DIALOG_BODY}>
          <H4>Welcome to your new {Constants.sourceAcademyDeploymentName} course!</H4>
          <div>
            Here at {Constants.sourceAcademyDeploymentName}, our mission is to bring out the beauty
            and fun in programming and the ideas behind programming, and to make these ideas
            universally accessible. This includes educational research!
          </div>
          <br />
          <div>
            We collect programs that students run in {Constants.sourceAcademyDeploymentName} and
            store them anonymously for our research. You are free to opt out of this collection,
            with no penalty for you whatsoever. Contact your course instructor if you have questions
            or concerns about this research.
          </div>
        </div>
        <div className={Classes.DIALOG_FOOTER}>
          <div className={Classes.DIALOG_FOOTER_ACTIONS}>
            <Button
              text="I would like to opt out"
              onClick={() => dispatch(updateCourseResearchAgreement(false))}
            />
            <Button
              text="I consent!"
              intent={Intent.SUCCESS}
              onClick={() => dispatch(updateCourseResearchAgreement(true))}
            />
          </div>
        </div>
      </Dialog>
    </div>
  );
};

export default ResearchAgreementPrompt;
