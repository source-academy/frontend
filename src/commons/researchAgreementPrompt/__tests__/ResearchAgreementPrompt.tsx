import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import { Provider } from 'react-redux';
import { useDispatch } from 'react-redux';
import { updateCourseResearchAgreement } from 'src/commons/application/actions/SessionActions';
import { mockInitialStore } from 'src/commons/mocks/StoreMocks';

import ResearchAgreementPrompt from '../ResearchAgreementPrompt';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn()
}));
const useDispatchMock = useDispatch as jest.Mock;
const dispatchMock = jest.fn();

const createElement = () => {
  const mockStore = mockInitialStore();
  const element = (
    <Provider store={mockStore}>
      <ResearchAgreementPrompt />
    </Provider>
  );

  return element;
};

describe('ResearchAgreementPrompt', () => {
  let user: UserEvent;

  beforeEach(() => {
    useDispatchMock.mockReturnValue(dispatchMock);
    dispatchMock.mockClear();
    user = userEvent.setup();
  });

  test('"I would like to opt out" dispatches correctly', async () => {
    render(createElement());
    const button = await screen.findByText('I would like to opt out');
    await user.click(button);

    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith(updateCourseResearchAgreement(false));
  });

  test('"I consent!" dispatches correctly', async () => {
    render(createElement());
    const button = await screen.findByText('I consent!');
    await user.click(button);

    expect(dispatchMock).toBeCalledTimes(1);
    expect(dispatchMock).toBeCalledWith(updateCourseResearchAgreement(true));
  });
});
