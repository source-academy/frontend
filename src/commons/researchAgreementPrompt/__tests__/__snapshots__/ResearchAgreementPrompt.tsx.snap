// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResearchAgreementPrompt renders correctly 1`] = `
"<Provider store={{...}}>
  <ResearchAgreementPrompt>
    <div className=\\"research-prompt\\">
      <Blueprint4.Dialog className=\\"bp4-dark\\" title=\\"Agreement to Participate in Educational Research\\" canOutsideClickClose={false} canEscapeKeyClose={false} isCloseButtonShown={false} isOpen={true}>
        <Blueprint4.Overlay className=\\"bp4-overlay-scroll-container\\" title=\\"Agreement to Participate in Educational Research\\" canOutsideClickClose={false} canEscapeKeyClose={false} isCloseButtonShown={false} isOpen={true} hasBackdrop={true} autoFocus={true} backdropProps={{...}} enforceFocus={true} lazy={true} shouldReturnFocusOnClose={true} transitionDuration={300} transitionName=\\"bp4-overlay\\" usePortal={true}>
          <Blueprint4.Portal className={[undefined]} container={{...}} stopPropagationEvents={[undefined]}>
            <Portal containerInfo={{...}}>
              <TransitionGroup appear={true} aria-live=\\"polite\\" className=\\"bp4-overlay bp4-overlay-open bp4-overlay-scroll-container\\" component=\\"div\\" onKeyDown={[Function (anonymous)]} childFactory={[Function: childFactory]}>
                <div aria-live=\\"polite\\" className=\\"bp4-overlay bp4-overlay-open bp4-overlay-scroll-container\\" onKeyDown={[Function (anonymous)]}>
                  <CSSTransition classNames=\\"bp4-overlay\\" addEndListener={[Function (anonymous)]} timeout={300} unmountOnExit={true} onExited={[Function: bound bound handleExited]} in={true} exit={[undefined]} enter={[undefined]}>
                    <Transition addEndListener={[Function (anonymous)]} timeout={300} unmountOnExit={true} onExited={[Function (anonymous)]} in={true} exit={true} enter={true} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExiting={[Function (anonymous)]} mountOnEnter={false} appear={false}>
                      <div tabIndex={0} className=\\"bp4-overlay-start-focus-trap\\" onFocus={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} />
                    </Transition>
                  </CSSTransition>
                  <CSSTransition classNames=\\"bp4-overlay\\" timeout={300} addEndListener={[Function (anonymous)]} onExited={[Function: bound bound handleExited]} in={true} exit={[undefined]} enter={[undefined]}>
                    <Transition timeout={300} addEndListener={[Function (anonymous)]} onExited={[Function (anonymous)]} in={true} exit={true} enter={true} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExiting={[Function (anonymous)]} mountOnEnter={false} unmountOnExit={false} appear={false}>
                      <div className=\\"bp4-overlay-backdrop\\" onMouseDown={[Function (anonymous)]} />
                    </Transition>
                  </CSSTransition>
                  <CSSTransition classNames=\\"bp4-overlay\\" onEntering={[undefined]} onEntered={[undefined]} onExiting={[undefined]} onExited={[Function: bound bound handleExited]} timeout={300} addEndListener={[Function (anonymous)]} in={true} exit={[undefined]} enter={[undefined]}>
                    <Transition onEntering={[Function (anonymous)]} onEntered={[Function (anonymous)]} onExiting={[Function (anonymous)]} onExited={[Function (anonymous)]} timeout={300} addEndListener={[Function (anonymous)]} in={true} exit={true} enter={true} onEnter={[Function (anonymous)]} onExit={[Function (anonymous)]} mountOnEnter={false} unmountOnExit={false} appear={false}>
                      <div className=\\"bp4-dialog-container bp4-overlay-content\\">
                        <div className=\\"bp4-dialog bp4-dark\\" role=\\"dialog\\" aria-labelledby=\\"title-bp-dialog-0\\" aria-describedby={[undefined]} style={[undefined]}>
                          <div className=\\"bp4-dialog-header\\">
                            <Blueprint4.Icon icon={[undefined]} size={16} aria-hidden={true} tabIndex={-1} />
                            <Component id=\\"title-bp-dialog-0\\">
                              <h6 id=\\"title-bp-dialog-0\\" className=\\"bp4-heading\\">
                                Agreement to Participate in Educational Research
                              </h6>
                            </Component>
                          </div>
                          <div className=\\"bp4-dialog-body\\">
                            <Component>
                              <h4 className=\\"bp4-heading\\">
                                Welcome to your new 
                                Source Academy @ NUS
                                 course!
                              </h4>
                            </Component>
                            <div>
                              Here at 
                              Source Academy @ NUS
                              , our mission is to bring out the beauty and fun in programming and the ideas behind programming, and to make these ideas universally accessible. This includes educational research!
                            </div>
                            <br />
                            <div>
                              We collect programs that students run in 
                              Source Academy @ NUS
                               and store them anonymously for our research. You are free to opt out of this collection, with no penalty for you whatsoever. Contact your course instructor if you have questions or concerns about this research.
                            </div>
                          </div>
                          <div className=\\"bp4-dialog-footer\\">
                            <div className=\\"bp4-dialog-footer-actions\\">
                              <Blueprint4.Button text=\\"I would like to opt out\\" onClick={[Function: onClick]}>
                                <button type=\\"button\\" onClick={[Function: onClick]} className=\\"bp4-button\\" disabled={false} onBlur={[Function (anonymous)]} onFocus={[undefined]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                  <Blueprint4.Icon icon={[undefined]} />
                                  <span className=\\"bp4-button-text\\">
                                    I would like to opt out
                                  </span>
                                  <Blueprint4.Icon icon={[undefined]} />
                                </button>
                              </Blueprint4.Button>
                              <Blueprint4.Button text=\\"I consent!\\" intent=\\"success\\" onClick={[Function: onClick]}>
                                <button type=\\"button\\" onClick={[Function: onClick]} className=\\"bp4-button bp4-intent-success\\" disabled={false} onBlur={[Function (anonymous)]} onFocus={[undefined]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                  <Blueprint4.Icon icon={[undefined]} />
                                  <span className=\\"bp4-button-text\\">
                                    I consent!
                                  </span>
                                  <Blueprint4.Icon icon={[undefined]} />
                                </button>
                              </Blueprint4.Button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </Transition>
                  </CSSTransition>
                  <CSSTransition classNames=\\"bp4-overlay\\" addEndListener={[Function (anonymous)]} timeout={300} unmountOnExit={true} onExited={[Function: bound bound handleExited]} in={true} exit={[undefined]} enter={[undefined]}>
                    <Transition addEndListener={[Function (anonymous)]} timeout={300} unmountOnExit={true} onExited={[Function (anonymous)]} in={true} exit={true} enter={true} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExiting={[Function (anonymous)]} mountOnEnter={false} appear={false}>
                      <div tabIndex={0} className=\\"bp4-overlay-end-focus-trap\\" onFocus={[Function (anonymous)]} />
                    </Transition>
                  </CSSTransition>
                </div>
              </TransitionGroup>
            </Portal>
          </Blueprint4.Portal>
        </Blueprint4.Overlay>
      </Blueprint4.Dialog>
    </div>
  </ResearchAgreementPrompt>
</Provider>"
`;
